src/Servers/IIS/AspNetCoreModuleV2/AspNetCore/applicationmanager.cpp

```diff
@@ -211,14 +218,19 @@ APPLICATION_MANAGER::RecycleApplicationFromManager(
 VOID
 APPLICATION_MANAGER::ShutDown()
 {
+    // During shutdown we lock until we delete the application
+    SRWExclusiveLock lock(m_srwLock);
+
     // We are guaranteed to only have one outstanding OnGlobalStopListening event at a time
     // However, it is possible to receive multiple OnGlobalStopListening events
     // Protect against this by checking if we already shut down.
+    if (g_fInShutdown)
```

Comment:
It sounds like this check wasn't sufficient?

Reply:
There wasn't a check before even though the comment mentioned there being a check. There is a check in `applicationInfo->ShutDownApplication` but that's a different class and I'd prefer to have a local check here, as well as fulfill the comments prophecy.

Reply:
I guess I assumed the check it was referring to was `if (g_fInShutdown)`

Reply:
Which is what I added?

Reply:
Wow.  The omission is so glaring I just unconsciously assumed it had been there all along. 🤦